src/olympia/addons/utils.py

```diff
@@ -247,3 +259,130 @@ def get_error_message(self):
+                    'protected DOM APIs: {permissions}.'
+                ).format(hostnames=hostnames, permissions=permissions),
+                'install_origins': self.install_origins,
+                'site_permissions': self.site_permissions,
+                'browser_specific_settings': {
+                    'gecko': {'strict_min_version': self.MIN_VERSION}
+                },
+            }
+        return manifest_data
+
+    def _create_zipfile(self, manifest_data):
+        # Create the xpi containing the manifest, in memory.
+        raw_buffer = BytesIO()
+        with zipfile.ZipFile(raw_buffer, 'w') as zip_file:
+            zip_file.writestr('manifest.json', json.dumps(manifest_data, indent=2))
+        raw_buffer.seek(0)
+        filename = 'automatic.xpi'
+        file_obj = File(raw_buffer, filename)
+
+        return file_obj
+
+    @atomic
+    def create_version(self, addon=None):
+        from olympia.addons.models import Addon
+        from olympia.versions.utils import get_next_version_number
+        from olympia.devhub.tasks import create_version_for_upload
+        from olympia.files.models import FileUpload
+        from olympia.files.utils import parse_addon
+
+        version_number = '1.0'
+
+        # If passing an existing add-on, we need to bump the version number
+        # to avoid clashes.
+        if addon is not None:
+            version_number = get_next_version_number(addon)
+
+        # Create the manifest, with more user-friendly name & description built
+        # from install_origins/site_permissions, and then the zipfile with that
+        # manifest inside.
+        manifest_data = self._create_manifest(version_number)
+        file_obj = self._create_zipfile(manifest_data)
+
+        # Parse the zip we just created. The user needs to be the Mozilla User
+        # because regular submissions of this type of add-on is forbidden to
+        # normal users.
+        parsed_data = parse_addon(
+            file_obj,
+            addon=addon,
+            user=UserProfile.objects.get(pk=settings.TASK_USER_ID),
+        )
+
+        if addon is None:
+            # Create the Addon instance (without a Version/File at this point).
+            addon = Addon.initialize_addon_from_upload(
+                data=parsed_data,
+                upload=file_obj,
+                channel=amo.RELEASE_CHANNEL_UNLISTED,
+                user=self.user,
+            )
+        else:
+            # If the user isn't an author, something is wrong.
+            if not addon.authors.filter(pk=self.user.pk).exists():
+                raise ImproperlyConfigured(
+                    'PermissionEnablerCreator was instantiated with a bogus addon/user'
+                )
+            # FIXME: if we already have an add-on, we'll likely want to update
+            # its name and description so that it matches the latest manifest
+            # values (which we have in parsed_data).
+
```

Comment:
Are we checking anywhere that the addon is already an `ADDON_PERMISSION_ENABLER` to prevent an addon ending up with a mix of types?  (I would hope that would raise elsewhere though)

Reply:
Good point, it doesn't hurt to throw an exception if the `type` is wrong